rtl/eclrtl/rtlkey.cpp

```diff
@@ -815,56 +826,88 @@ class CIndirectKeySegmentMonitor : implements IKeySegmentMonitor, public CInterf
     unsigned offset;
 };
 
-//The base monitor provided to this segment monitor is constructed with offsets of 0
-class CVarOffsetKeySegmentMonitor : public CIndirectKeySegmentMonitor
+// The base monitor provided to this segment monitor is constructed with offsets of 0
+// offset refers to where the field would be in an "expanded" version of the record (all variable size - and thus unkeyed - fields being assumed null)
+
+class CNewVarOffsetKeySegmentMonitor : public CIndirectKeySegmentMonitor
 {
 public:
-    CVarOffsetKeySegmentMonitor(IKeySegmentMonitor * _base, unsigned _offset, IKeySegmentOffsetTranslator * _offsetTranslator) : CIndirectKeySegmentMonitor(_base, _offset) { offsetTranslator.setown(_offsetTranslator); }
+    CNewVarOffsetKeySegmentMonitor(IKeySegmentMonitor * _base, unsigned _offset, unsigned _fieldIdx)
+    : CIndirectKeySegmentMonitor(_base, _offset), fieldIdx(_fieldIdx)
+    {
+    }
 
-    CVarOffsetKeySegmentMonitor(MemoryBuffer &mb)
-        : CIndirectKeySegmentMonitor(mb)
+    CNewVarOffsetKeySegmentMonitor(MemoryBuffer &mb)
+    : CIndirectKeySegmentMonitor(mb)
     {
-        throwUnexpected();
+        mb.read(fieldIdx);
+    }
+
+    virtual bool matches(const RtlRow *keyval) const
+    {
+        return base->matchesBuffer(getSegmentBase(keyval));
     }
 
-    virtual bool matches(const void *keyval) const
+    virtual unsigned queryFieldNo() const { return fieldIdx+1; }
+
+    // We can't presently build in-memory indexes for varoffset fields, as we don't have a way to
+    // translate the offsets in the lhs (keybuffer) side of the compare.
+    // I suppose if we wanted to lift that we could pass in RtlRow for both parameters?
+    // For now, these methods can never be called
+
+    virtual bool matchesBuffer(const void *keyval) const
     {
-        return base->matches(offsetTranslator->getSegmentBase(keyval));
+        throwUnexpected();
     }
     virtual int docompare(const void * expandedLeft, const void * rawRight) const
     {
-        return base->docompare((const byte *)expandedLeft + offset, offsetTranslator->getSegmentBase(rawRight));
+        throwUnexpected();
     }
     virtual void copy(void * expandedRow, const void * rawRight) const
     {
-        base->copy((byte *)expandedRow + offset, offsetTranslator->getSegmentBase(rawRight));
+        throwUnexpected();
     }
     virtual int docompareraw(const void * left, const void * right) const
     {
-        return base->docompare(offsetTranslator->getSegmentBase(left), offsetTranslator->getSegmentBase(right));
+        throwUnexpected();
     }
+
```

Comment:
Should you also fail on various other functions - increment, setLow, setHigh, or implement them?